### PR TITLE
arrow: introduce Arrow wrapper

### LINKED ISSUE #663: Update fsspec's HDFS filesystem to use the new pyarrow.fs.HadoopFileSystem
See https://github.com/intake/filesystem_spec/issues/295#issuecomment-669764372 for context. 

The current implementation included in `fsspec` for a HDFS filesystem relies on `pyarrow.hdfs.HadoopFileSystem`, which is deprecated in favor of `pyarrow.fs.HadoopFileSystem`. 

In general, this will require implementing an fsspec-compatible wrapper for the new arrow filesystems (so a wrapper "the other way around" as the one for fsspec-filesystems we have in pyarrow). Such a general wrapper could then also be used for the hdfs wrapper.

Some time ago I started such a wrapper, which could be used as a starting point: https://github.com/intake/filesystem_spec/compare/master...jorisvandenbossche:pyarrow-wrapper?expand=1

### PR DESCRIPTION
**Pull request by isidentical**
Resolves #663 (based on the initial work done by @jorisvandenbossche).

TODO:
- [x] Add tests
- [x] Docs / Registration

### CODE REVIEW

File: fsspec/implementations/arrow.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+            if not exception.args:
+                raise
+
+            message, *args = exception.args
+            if isinstance(message, str) and "does not exist" in message:
+                raise FileNotFoundError(errno.ENOENT, message) from exception
+            else:
+                raise
+
+    return wrapper
```
Comment (efiop) [inline fsspec/implementations/arrow.py:28]: For the record: https://github.com/intake/filesystem_spec/pull/749#discussion_r706487444

Comment (isidentical) [reply fsspec/implementations/arrow.py:28]: Unfortunately they are still present, without this many places I see just raise bare OSError s

Comment (jorisvandenbossche) [reply fsspec/implementations/arrow.py:28]: Feel free to open a JIRA on the Arrow side about this, if you still run into example cases that raise bare OSError for such cases.

File: fsspec/implementations/arrow.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        self.kwargs = kwargs
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        return self.close()
+
+
+class HadoopFileSystemWrapper(ArrowFSWrapper):
```
Comment (jorisvandenbossche) [inline fsspec/implementations/arrow.py:190]: Maybe for this class you can leave off the "Wrapper" in the name? As for the end-user, they are not explicitly wrapping another filesystem (as you do with `ArrowFSWrapper`), this is only happening as an internal implementation detail.

File: fsspec/implementations/tests/test_arrow.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    fs.move(remote_dir + "/a.txt", remote_dir + "/b.txt")
+    secondary_info = fs.info(remote_dir + "/b.txt")
+
+    assert not fs.exists(remote_dir + "/a.txt")
+    assert fs.exists(remote_dir + "/b.txt")
+
+    initial_info.pop("name")
+    secondary_info.pop("name")
+    assert initial_info == secondary_info
```
Comment (jorisvandenbossche) [inline fsspec/implementations/tests/test_arrow.py:59]: You might also want to add a test moving to a directory (with recursive=True), see eg https://github.com/intake/filesystem_spec/blob/0589358d8a029ed6b60d031018f52be2eb721291/fsspec/implementations/tests/test_local.py#L622

**Review by jorisvandenbossche — COMMENTED** [review]
Cool, thanks a lot for picking up my initial implementation!

File: fsspec/implementations/arrow.py (hunk 2 above)
Comment (isidentical) [reply fsspec/implementations/arrow.py]: Makes sense, changed!

File: fsspec/implementations/tests/test_arrow.py (hunk 3 above)
Comment (isidentical) [reply fsspec/implementations/tests/test_arrow.py:59]: Sure, done!

File: fsspec/implementations/arrow.py (hunk 1 above)
Comment (isidentical) [reply fsspec/implementations/arrow.py:28]: Here is a simple example, if I can get some time I will try to create a ticker for this and other scenarios I've experienced.
```terminal
(.venv38) (Python 3.8.5+) [  5:51ÖS ]  [ <hey>@<hey>:~/filesystem_spec(pyarrow✗) ]
 $ pip show pyarrow | grep -i version
Version: 4.0.0
License: Apache License, Version 2.0
(.venv38) (Python 3.8.5+) [  5:51ÖS ]  [ <hey>@<hey>:~/filesystem_spec(pyarrow✗) ]
 $ pytest fsspec/implementations/tests/test_arrow.py
============================================================================== test session starts ===============================================================================
platform linux -- Python 3.8.5+, pytest-6.2.5, py-1.9.0, pluggy-0.13.1
rootdir: /home/isidentical/filesystem_spec
plugins: timeout-1.4.2, mock-3.6.1, lazy-fixture-0.6.3, anyio-2.0.2, tap-3.2, env-0.6.2, cov-2.12.1, flaky-3.7.0, xdist-2.3.0, docker-0.10.3, apiritif-0.9.8, forked-1.3.0, asyncio-0.14.0
collected 11 items                                                                                                                                                               

fsspec/implementations/tests/test_arrow.py ........F..                                                                                                                     [100%]

==================================================================================== FAILURES ====================================================================================
________________________________________________________________________________ test_exceptions _________________________________________________________________________________

fs = <fsspec.implementations.arrow.ArrowFSWrapper object at 0x7f860c1810a0>, remote_dir = 'f41760ad91b75cc4ace4e99a94c7e9a8'

    def test_exceptions(fs, remote_dir):
        with pytest.raises(FileNotFoundError):
>           with fs.open(remote_dir + "/a.txt"):

fsspec/implementations/tests/test_arrow.py:156: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
fsspec/spec.py:978: in open
    f = self._open(
fsspec/implementations/arrow.py:144: in _open
    stream = self.fs.open_input_stream(path)
pyarrow/_fs.pyx:621: in pyarrow._fs.FileSystem.open_input_stream
    ???
pyarrow/error.pxi:141: in pyarrow.lib.pyarrow_internal_check_status
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   OSError: Path does not exist 'f41760ad91b75cc4ace4e99a94c7e9a8/a.txt'

pyarrow/error.pxi:112: OSError
============================================================================ short test summary info =============================================================================
FAILED fsspec/implementations/tests/test_arrow.py::test_exceptions - OSError: Path does not exist 'f41760ad91b75cc4ace4e99a94c7e9a8/a.txt'
========================================================================== 1 failed, 10 passed in 0.21s ==========================================================================
```

File: docs/source/api.rst [hunk 4]
```diff
@@ -97,6 +97,8 @@ Built-in Implementations
 .. autosummary::
    fsspec.implementations.ftp.FTPFileSystem
    fsspec.implementations.hdfs.PyArrowHDFS
+   fsspec.implementations.arrow.ArrowFSWrapper
+   fsspec.implementations.arrow.HadoopFileSystem
```
Comment (martindurant) [inline docs/source/api.rst:101]: also needs autoclass entries below

File: fsspec/utils.py [hunk 5]
```diff
@@ -447,3 +448,21 @@ def setup_logging(logger=None, logger_name=None, level="DEBUG", clear=True):
     logger.addHandler(handle)
     logger.setLevel(level)
     return logger
+
+
+def mirror_from(origin_name, methods):
```
Comment (martindurant) [inline fsspec/utils.py:453]: This could probably use a test

**Comment by martindurant** [comment]
Thanks for reviewing @jorisvandenbossche - you can have the final OK on this one.

File: setup.py [hunk 6]
```diff
@@ -45,6 +45,7 @@
         "github": ["requests"],
         "gs": ["gcsfs"],
         "hdfs": ["pyarrow >= 1"],
+        "arrow": ["pyarrow >= 1"],
```
[the remark this answers is not in the document]
Comment (isidentical) [reply setup.py:48]: well technically it is the `ArrowWrapper` what are you are getting, so unsure about this but also makes sense if `HadoopFileSystem` is all we are going to export from it.

File: fsspec/utils.py (hunk 5 above)
Comment (isidentical) [reply fsspec/utils.py:453]: Done!

**Comment by jorisvandenbossche** [comment]
Thanks @isidentical!
I updated the issue on the Arrow side about this: https://issues.apache.org/jira/browse/ARROW-7102. And once this is released, can mention this in the pyarrow docs as well.

**Pull request merged**